geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryMonitor.java

```diff
@@ -161,51 +161,26 @@ public void run() {
     // if the query monitor is stopped before run has been called, we should not run
     synchronized (this.stopped) {
       if (this.stopped.get()) {
-        queryThreads.clear();
+        queryExpirations.clear();
         return;
       }
       this.monitoringThread = Thread.currentThread();
     }
     try {
-      QueryThreadTask queryTask;
-      long sleepTime;
       while (true) {
-        // Get the first query task from the queue. This query will have the shortest
-        // remaining time that needs to canceled first.
-        queryTask = (QueryThreadTask) queryThreads.peek();
-        if (queryTask == null) {
-          synchronized (queryThreads) {
-            queryThreads.wait();
-          }
-          continue;
-        }
+        final QueryExpiration queryExpiration = queryExpirations.take();
```

Comment:
This take could slow down a bit since it will now crawl the entire queue until it finds one that is expired.  I don't know if this will actually matter overall.

Reply:
This shouldn't be an issue - the `Delayed` interface on the `QueryExpiration` object implements `compareTo`, which will put the oldest candidates (most eligible for expiration) at the front of the queue.